extensions/amp-timeago/0.2/timeago.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createElement, useEffect, useRef, useState} from '../../../src/preact';
+import {timeago} from '../../../third_party/timeagojs/timeago';
+import {useResourcesNotify} from '../../../src/preact/utils';
+
+/**
+ * @param {!JsonObject} props
+ * @return {Preact.Renderable}
+ */
+export function Timeago(props) {
+  const {
+    'datetime': datetime,
+    'locale': locale,
+    'cutoff': cutoff,
+    'cutoffText': cutoffText,
+  } = props;
+  const [timestamp, setTimestamp] = useState(null);
+  const ref = useRef(null);
+
+  useEffect(() => {
+    const node = ref.current;
+    const observer = new IntersectionObserver(entries => {
+      const last = entries[entries.length - 1];
+      if (last.isIntersecting) {
+        setTimestamp(
+          getFuzzyTimestampValue(datetime, locale, cutoffText, cutoff)
+        );
+      }
+    });
+    if (node) {
+      observer.observe(node);
+    }
+    return () => observer.disconnect();
+  }, [datetime, locale, cutoff, cutoffText]);
```

Comment:
Feel free to ignore since everyone else has approved this, but won't this create a new IntersectionObserver on any prop change?  (Thats why I was thinking this should be a different `useEffect`)

Reply:
Yes it will. We discussed as an optimization, we can separate side-effects like you suggested or use one side-effect + `useMemo` to re-calculate the value. E.g.

```
const [intersectionCount, setIntersectionCount] = useState(0);
const timestamp = useMemo(() => {
  if (intersectionCount < 1) {
    return null;
  }
  return getFuzzyTimestamp(datetime, locale, cutoff, cutoffText);
}, [intersectionCount, datetime, locale, cutoff, cutoffText]);

useEffect(() => {
  new IntersectionObserver(() => {
    if (isIntersecting) {
      setIntersectionCount(old => old + 1);
    }
  })
}, []);

return <div>{timestamp}</div>
```

I think this would work, and likely a better option vs current state. It's just that we wanted to merge the first version and continue discussion with the merged code.

Reply:
Cool. Thanks for the explanation, makes sense to me 👍